messaging/delivery/http/chat_delivery.go

```diff
@@ -0,0 +1,191 @@
+	broadcast:  make(chan domain.Message),
+	Register:   make(chan subscription),
+	unregister: make(chan subscription),
+	rooms:      make(map[string]map[subscription]bool),
+}
+
+const (
+	maxMessageSize int64 = 1024
+	pongWait             = time.Minute * 5
+	pingPeriod           = time.Minute * 5
+	writeWait            = time.Minute
+)
+
+func (s *subscription) readPump(u domain.MessageUseCase) {
+	c := s.conn
+	defer func() {
+		MainHub.unregister <- *s
+		c.ws.Close()
+	}()
+	c.ws.SetReadLimit(maxMessageSize)
+	_ = c.ws.SetReadDeadline(time.Now().Add(pongWait))
+	c.ws.SetPongHandler(func(string) error { _ = c.ws.SetReadDeadline(time.Now().Add(pongWait)); return nil })
+	for {
+		_, msg, err := c.ws.ReadMessage()
+		if err != nil {
+			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway) {
+				log.Printf("error: %v", err)
+			}
+			break
+		}
+		m := domain.Message{RoomID: s.roomID, SentTimestamp: time.Now(), FromStudentID: s.userID, MessageBody: string(msg)}
+		err = u.SendMessage(context.Background(), &m)
+		if err != nil {
+			log.Printf("Failed to save message with err %s", err.Error())
+		}
+		MainHub.broadcast <- m
+	}
+}
+
+func (s *subscription) writePump() {
+	c := s.conn
+	ticker := time.NewTicker(pingPeriod)
+	defer func() {
+		ticker.Stop()
+		c.ws.Close()
+	}()
+	for {
+		select {
+		case message, ok := <-c.send:
+			if !ok {
+				c.write(websocket.CloseMessage, []byte{})
+				return
+			}
+			res, err := json.Marshal(message)
+			if err != nil {
+				log.Printf("message %s couldn't be sent to %s in room %s.", message, s.userID, s.roomID)
+			}
+			if err = c.write(websocket.TextMessage, res); err != nil {
+				log.Printf("message %s couldn't be sent to %s in room %s.", message, s.userID, s.roomID)
+				return
+			}
+		case <-ticker.C:
+			if err := c.write(websocket.PingMessage, []byte{}); err != nil {
+				return
+			}
+		}
+	}
+}
+
+func (c *connection) write(mt int, payload []byte) error {
+	c.ws.SetWriteDeadline(time.Now().Add(writeWait))
+	return c.ws.WriteMessage(mt, payload)
+}
+
+// ServeWs is the handleFunc for connecting to a room's websocket. A user must be authorized, i.e. already added to
+// the room before he can connect to the room. Otherwise, returns 401.
+func (h *MessageHandler) ServeWs(w http.ResponseWriter, r *http.Request, roomID string, userID string) {
+	authorized := h.u.IsAuthorized(userID, roomID)
+	if !authorized {
+		w.WriteHeader(http.StatusUnauthorized)
+		io.WriteString(w, "Not authorized to enter the room number "+roomID)
+		return
+	}
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Println(err.Error())
+		return
+	}
+	c := &connection{send: make(chan domain.Message), ws: ws}
+	s := subscription{c, roomID, userID}
+	if authorized {
+		MainHub.Register <- s
+	}
+	go s.writePump()
+	go s.readPump(h.u)
+}
+
+// StartHeartbeat starts the functioning of the Hub with respect to registering, unregistering and broadcasting
+func (h *Hub) StartHeartbeat() {
```

Comment:
rename method: I thought this was a heartbeat mechanism based on the name, but it looks moreso like a listener, can we rename this method to maybe StartHubListener?